Sources/SourceKitLSP/WorkDoneProgressManager.swift

```diff
@@ -177,3 +190,68 @@ final actor WorkDoneProgressManager {
     }
   }
 }
+
+/// A `WorkDoneProgressManager` that essentially has two states. If any operation tracked by this type is currently
+/// running, it displays a work done progress in the client. If multiple operations are running at the same time, it
+/// doesn't show multiple work done progress in the client. For example, we only want to show one progress indicator
+/// when sourcekitd has crashed, not one per `SwiftLanguageService`.
+actor SharedWorkDoneProgressManager {
+  private weak var sourceKitLSPServer: SourceKitLSPServer?
+
+  /// The number of in-progress operations. When greater than 0 `workDoneProgress` non-nil and a work done progress is
+  /// displayed to the user.
+  private var inProgressOperations = 0
+  private var workDoneProgress: WorkDoneProgressManager?
+
+  private let tokenPrefix: String
+  private let title: String
+  private let message: String?
+
+  public init(
+    sourceKitLSPServer: SourceKitLSPServer,
+    tokenPrefix: String,
+    title: String,
+    message: String? = nil
+  ) {
+    self.sourceKitLSPServer = sourceKitLSPServer
+    self.tokenPrefix = tokenPrefix
+    self.title = title
+    self.message = message
+  }
+
+  func start() async {
+    guard let sourceKitLSPServer else {
+      return
+    }
+    // Do all asynchronous operations up-front so that incrementing `inProgressOperations` and setting `workDoneProgress`
+    // cannot be interrupted by an `await` call
+    let initialDebounce = await sourceKitLSPServer.options.workDoneProgressDebounceDuration
+    let capabilityRegistry = await sourceKitLSPServer.capabilityRegistry
+
+    inProgressOperations += 1
+    if let capabilityRegistry, workDoneProgress == nil {
+      workDoneProgress = WorkDoneProgressManager(
+        server: sourceKitLSPServer,
+        capabilityRegistry: capabilityRegistry,
+        tokenPrefix: tokenPrefix,
+        initialDebounce: initialDebounce,
+        title: title,
+        message: message
+      )
+    }
+  }
+
+  func end() async {
+    if inProgressOperations > 0 {
+      inProgressOperations -= 1
+    } else {
+      logger.fault(
+        "Unbalanced calls to SharedWorkDoneProgressManager.start and end for \(tokenPrefix, privacy: .public)"
```

Comment:
I checked this out locally but the patch failed to build because `tokenPrefix` should be `self.tokenPrefix`. After fixing this it worked as expected.

```
sourcekit-lsp/Sources/SourceKitLSP/WorkDoneProgressManager.swift:249:80: error: reference to property 'tokenPrefix' in closure requires explicit use of 'self' to make capture semantics explicit
247 |     } else {
248 |       logger.fault(
249 |         "Unbalanced calls to SharedWorkDoneProgressManager.start and end for \(tokenPrefix, privacy: .public)"
    |                                                                                |- error: reference to property 'tokenPrefix' in closure requires explicit use of 'self' to make capture semantics explicit
    |                                                                                `- note: reference 'self.' explicitly
250 |       )
251 |     }
```